packages/commerce-sdk-react/src/auth/index.ts

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright (c) 2022, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import {helpers, ShopperLogin, ShopperLoginTypes} from 'commerce-sdk-isomorphic'
+import jwtDecode from 'jwt-decode'
+import {ApiClientConfigParams} from '../hooks/types'
+import {BaseStorage, LocalStorage, CookieStorage} from './storage'
+
+type Helpers = typeof helpers
+interface AuthConfig extends ApiClientConfigParams {
+    redirectURI: string
+    proxy: string
+}
+
+interface JWTHeaders {
+    exp: number
+    iat: number
+}
+
+// this type is slightly different from ShopperLoginTypes.TokenResponse, reasons:
+// 1. TokenResponse is too generic (with & {[key:string]: any}), we need a more
+//    restrictive type to make sure type safe
+// 2. The refresh tokens are stored separately for guest and registered user. Instead
+//    of refresh_token, we have refresh_token_guest and refresh_token_registered
+type AuthDataKeys =
+    | 'access_token'
+    | 'customer_id'
+    | 'enc_user_id'
+    | 'expires_in'
+    | 'id_token'
+    | 'idp_access_token'
+    | 'refresh_token_guest'
+    | 'refresh_token_registered'
+    | 'token_type'
+    | 'usid'
+type AuthDataMap = Record<
+    AuthDataKeys,
+    {
+        storage: BaseStorage
+        key: string
+        callback?: () => void
+    }
+>
+
+/**
+ * This class is used to handle shopper authentication.
+ * It is responsible for initializing shopper session, manage access
+ * and refresh tokens on server/browser environments. As well as providing
+ * a mechanism to queue network calls before having a valid access token.
+ *
+ * @Internal
+ */
+class Auth {
```

Comment:
I like what you've done with `pending` to queue the data, but there's a bit that you could DRY. Mostly you have situations like this:
```ts
async method() {
  const request = () => {
    // Method-specific logic, then
    const res = await login() // Makes request
    this.handleTokenResponse(res, true) // Saves `res` to storage
    return this.data // Reconstructs `res` from storage
  }
  this.pending = this.pending ? this.pending.then(request) : request()
  return this.pending
```
(Also, it seems circuitous to save `res` to storage in `handleTokenResponse` and then re-construct it from storage with the `data` getter. We already have it pre-constructed!)

Abstract it into a private helper method, then the public methods (including `init`) can be simplified!
```ts
{
    async queueRequest(fn: () => Promise<ShopperLoginTypes.TokenResponse>, isGuest: boolean) {
        const queue = this.pending ?? Promise.resolve()
        this.pending = queue.then(async () => {
            const token = await fn()
            // This is the only place `handleTokenResponse` is used, so you could even pull it in to this method!
            this.handleTokenResponse(token, isGuest)
            return token
        })
        return this.pending
    }
    async logout() {
      return this.queueRequest(() => { /* ... */ }, true)
    }
}
```

Reply:
✅ Good suggestion and I've updated the code.

> Also, it seems circuitous to save res to storage in handleTokenResponse and then re-construct it from storage with the data getter. We already have it pre-constructed!

When user opens multiple tabs, the data in memory could be outdated and we must grab the data from storage (cookie/localstorage) directly